# Using STARsolo to generate count matrices from raw fastq files
## *Input*


**RNAseq reads**

- Tar files containing the raw fastq reads (e.g., .fastq.tar) have to be extracted first, using `tar -xvf *.tar`
- Each directory extracted from the tar files should contain `fastq.gz` files for R1 and R2 respectively (e.g., `R1.fastq.gz` and `R2.fastq.gz`)

**Genome Directory**

- Generated by STAR, from FASTA file and GTF annotation file (see https://physiology.med.cornell.edu/faculty/skrabanek/lab/angsd/lecture_notes/STARmanual.pdf)

**File manifest**

Given `directory_list`, a Python list of absolute paths to the directories extracted from `.fastq.tar` files, the file manifest can be generated by the code below:
```python {cmd}
file_manifest = "\n".join(
    [(lambda directory: "\t".join([str(file) for file in Path(directory).iterdir()][::-1] +
    [Path(directory).name]))(directory_path) for directory_path in directory_list]
    )
with open("file_manifest.tsv", 'w') as f:
    f.write(file_manifest)

```
The `directory_list` can be generated using the bash command below:
```bash {cmd}
ls -d /path/to/the/directory/that/contains/the/extracted/cell/directories/*/ \
| egrep '[ATGC]{8,}/$' \ #to select for valid directory names
> directory_list.txt
```

## *Running STAR*
Make sure you are using STAR 2.7.10b (not available on the SCC), to install STAR 2.7.10b in your conda environment:
```bash {cmd}
conda install -c bioconda star
```
An example of a STARsolo command with parameters suitable for generating count matrices for downstream analyses in scQuint or Scanpy.

```bash {cmd}
STAR --runThreadN 12 \
--genomeDir /external/rprshnas01/netdata_kcni/stlab/Genomic_references/Ensembl/Mouse/Release_104/USE_THIS_genomeDir/ \
--readFilesCommand zcat \
--soloType SmartSeq \
--soloUMIdedup Exact \
--readFilesManifest file_manifest.tsv \
--soloOutFileNames output/ features.tsv barcodes.tsv matrix.mtx \
--soloStrand Unstranded \
--soloFeatures Gene SJ
```
Submitting the script as a Slurm job using qbatch. Running STARsolo on all samples in the Gouwens dataset (more than 4000 cells) took around 43 hours.
```bash {cmd}
qbatch -w 48:00:00 --ppj 12 --mem 60G -- bash STARsolo.sh
```